microsoft/atlas-design
# create build for tokens

task-378375

This PR creates a new package in the repo for the express purpose of taking our token files and transfering them into json and js format. This PR also includes a small proof of concept of how tokens can be integrated into the `/site`'s build system.

## Testing

1. Checkout branch.
2. Yarn install.
3. See below.

### Command to test

- `yarn build tokens` - runs the code to read our scss and copy it into json and js. You should see the output in the `/plugins/tokens/dist` folder.
- `yarn start` - should work and you should be able to view the typography page with tokens interpolated (though it isn't pretty yet!)
- after the above visit: http://localhost:1111/tokens/typography.html
- `yarn build` - should also work.


## Additional information

This package relies on `sass-extract` for it's parsing of Scss. Unfortunately `Sass-extract` relies on `node-sass` (Atlas will continue to use dart-sass). In order to cirumvent problems, it was necessary to include the tokens code into a package separate from /css.

The implications of this are as follows:
1.) The syntax for the `/css/tokens` folder must use `@import` as node-sass does not support @forward or @use.
2.) We cannot easily use this package to further tokenize files outside the /tokens folder.

Additionally, the code `/plugins/tokens` makes some changes to the structure returned by `extract-sass`. The reason for this is to provide great interop. Mustache for example, doesn't accessing/displaying keys in objects, so I've included both a `.list ([])` and `.map ({})` property on tokens. This means we'll get the quick access of variables with `.map` as well easy iteration with `.list`. It also means the files are larger than they would otherwise be.
